test/e2e/cluster_extension_install_test.go

```diff
@@ -593,6 +598,90 @@ func TestClusterExtensionInstallReResolvesWhenManagedContentChanged(t *testing.T
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: "default",
+		},
+	}
+	err := c.Create(context.Background(), sa)
+	require.NoError(t, err)
+
+	defer testCleanup(t, extensionCatalog, clusterExtension, sa)
+	defer getArtifactsOutput(t)
+
+	clusterExtension.Spec = ocv1alpha1.ClusterExtensionSpec{
+		PackageName:      "prometheus",
+		InstallNamespace: "default",
+		ServiceAccount: ocv1alpha1.ServiceAccountReference{
+			Name: sa.Name,
+		},
+	}
+	t.Log("It resolves the specified package with correct bundle path")
+	t.Log("By creating the ClusterExtension resource")
+	require.NoError(t, c.Create(context.Background(), clusterExtension))
+
+	t.Log("By eventually reporting a successful resolution and bundle path")
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		assert.NoError(ct, c.Get(context.Background(), types.NamespacedName{Name: clusterExtension.Name}, clusterExtension))
+		assert.Len(ct, clusterExtension.Status.Conditions, len(conditionsets.ConditionTypes))
+		cond := apimeta.FindStatusCondition(clusterExtension.Status.Conditions, ocv1alpha1.TypeResolved)
+		if !assert.NotNil(ct, cond) {
+			return
+		}
+		assert.Equal(ct, metav1.ConditionTrue, cond.Status)
+		assert.Equal(ct, ocv1alpha1.ReasonSuccess, cond.Reason)
+		assert.Contains(ct, cond.Message, "resolved to")
+		assert.Equal(ct, &ocv1alpha1.BundleMetadata{Name: "prometheus-operator.1.2.0", Version: "1.2.0"}, clusterExtension.Status.ResolvedBundle)
+	}, pollDuration, pollInterval)
+
+	t.Log("By eventually reporting a successful unpacked")
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		assert.NoError(ct, c.Get(context.Background(), types.NamespacedName{Name: clusterExtension.Name}, clusterExtension))
+		cond := apimeta.FindStatusCondition(clusterExtension.Status.Conditions, ocv1alpha1.TypeUnpacked)
+		if !assert.NotNil(ct, cond) {
+			return
+		}
+		assert.Equal(ct, metav1.ConditionTrue, cond.Status)
+		assert.Equal(ct, ocv1alpha1.ReasonUnpackSuccess, cond.Reason)
+		assert.Contains(ct, cond.Message, "unpack successful")
+	}, pollDuration, pollInterval)
+
+	t.Log("By eventually failing to install the package successfully due to insufficient ServiceAccount permissions")
+	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		assert.NoError(ct, c.Get(context.Background(), types.NamespacedName{Name: clusterExtension.Name}, clusterExtension))
+		cond := apimeta.FindStatusCondition(clusterExtension.Status.Conditions, ocv1alpha1.TypeInstalled)
+		if !assert.NotNil(ct, cond) {
+			return
+		}
+		assert.Equal(ct, metav1.ConditionFalse, cond.Status)
+		assert.Equal(ct, ocv1alpha1.ReasonInstallationFailed, cond.Reason)
+		assert.Contains(ct, cond.Message, "forbidden")
+	}, pollDuration, pollInterval)
+
+	t.Log("By fixing the ServiceAccount permissions")
+	require.NoError(t, createClusterRoleAndBindingForSA(context.Background(), name, sa))
+
+	t.Log("By eventually installing the package successfully")
```

Comment:
I assume we can't actually watch for the fixed service account, so this is predicated on exponential backoff retries?  If so, we should:

1. Be very careful to avoid touching the ClusterExtension after creating the necessary permissions (maybe we add a comment here about that)
2. Add a comment in the reconciler at the `return` site of the permission error to make sure that we don't swallow it and return a `nil` error.

Reply:
Correct. I am happy to add comments in places we feel necessary, but the comment about the permissions error in the reconciler is a bit iffy for me. The permissions error is being returned by the Helm Applier implementation, which gets the error from the Helm client. Maybe adding a comment saying "we should never eat errors returned from X"?

Reply:
> Maybe adding a comment saying "we should never eat errors returned from X"?

Yeah, that's what I had in mind generally. I think we should explain _why_ we are not going to eat those errors. That way our future selves are reminded about this specific thing if we ever decide we _do_ want to eat that error for some reason. In that kind of situation, it would help us remember that we would need to check the error to decide whether to eat it or not.

Reply:
Comments added in appropriate places in https://github.com/operator-framework/operator-controller/pull/1104/commits/a1d5ead220d75fe44fb5f1e17f7a6c62bbf16351